### PR TITLE
Add change membership to teams package

### PR DESCRIPTION
**Pull request by patrickxb**

### CODE REVIEW

**Comment by maxtaco** [comment]
Sorry for any weirdness in this review, github was acting super flakey at the time.

File: go/teams/box.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	Generation      int
+	Ctext           string
+	PerUserKeySeqno keybase1.Seqno  `json:"per_user_key_seqno"`
+	ReaderKeyMasks  []ReaderKeyMask `json:"reader_key_masks"`
+}
+
+type ReaderKeyMask struct {
+	Application int
+	Generation  int
+	Mask        string
```
Comment (patrickxb) [inline go/teams/box.go]: The overall ReaderKeyMask has a function to base64 decode.

File: go/teams/get.go [hunk 2]
```diff
@@ -40,22 +48,28 @@ func (f *finder) find(ctx context.Context, name string) (*TeamSigChainState, err
...
+func (f *finder) rawTeam(ctx context.Context, name string) (*rawTeam, error) {
 	arg := libkb.NewRetryAPIArg("team/get")
 	arg.NetContext = ctx
 	arg.SessionType = libkb.APISessionTypeREQUIRED
 	arg.Args = libkb.HTTPArgs{
 		"name": libkb.S{Val: name},
 	}
-	var chain rawChain
-	if err := f.G().API.GetDecode(arg, &chain); err != nil {
+	var rt rawTeam
```
Comment (patrickxb) [inline go/teams/get.go]: My mistake, it should have that field.

File: go/teams/box.go (hunk 1 above)
Comment (maxtaco) [inline go/teams/box.go]: It should still decode into a type specific for masks.

File: go/teams/box.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	case keybase1.TeamApplication_SALTPACK:
+		derivationString = libkb.TeamSaltpackDerivationString
+	}
+
+	maskBytes, err := mask.MaskBytes()
+	if err != nil {
+		return libkb.NaclDHKeyPair{}, err
+	}
+	var secBytes []byte
+	n := libkb.XORBytes(secBytes, derivedSecret(secret, derivationString), maskBytes)
```
Comment (maxtaco) [inline go/teams/box.go]: Can you make a typesafe wrapper here? I'm thinking back to bug 3964 or whatever...

**Comment by patrickxb** [comment]
As far as the round-trips, as I mentioned in kb chat, *everything* with teams, puks hits the network all the time.

I wanted to get this in first so integrations can start to happen and then do optimizations later.

**Comment by maxtaco** [comment]
I know, I'm just concerned we're going to triple our efforts down the road by ignoring it upfront.

File: go/teams/member_set.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+		return err
+	}
+	return nil
+}
+
+func (m *memberSet) loadGroup(ctx context.Context, g *libkb.GlobalContext, group []string) ([]member, error) {
+	members := make([]member, len(group))
+	var err error
+	for i, username := range group {
+		members[i], err = m.loadMember(ctx, g, username)
```
Comment (patrickxb) [inline go/teams/member_set.go]: At the time I started, I don't think UPAK had the necessary fields, but looks like they do now.

File: go/teams/box.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	"fmt"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+// TeamBox comes from api server team/get endpoint.
+type TeamBox struct {
+	Nonce           string
+	SenderKID       keybase1.KID `json:"sender_kid"`
```
Comment (patrickxb) [inline go/teams/box.go]: Ok.

File: go/teams/box.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	Generation      int
+	Ctext           string
+	PerUserKeySeqno keybase1.Seqno  `json:"per_user_key_seqno"`
+	ReaderKeyMasks  []ReaderKeyMask `json:"reader_key_masks"`
+}
+
+type ReaderKeyMask struct {
+	Application int
```
Comment (patrickxb) [inline go/teams/box.go]: To unmarshal JSON into a struct, Unmarshal matches incoming object keys to the keys used by Marshal (either the struct field name or its tag), preferring an exact match but also accepting a case-insensitive match. Unmarshal will only set exported fields of the struct.

**Comment by maxtaco** [comment]
I agree this iteration doesn't need to be fast, but I don't want us left with a 20 point task to fix all of the performance mistake we made earlier.

**Comment by patrickxb** [comment]
I agree.  I will make the UPAK changes here.

But (for example) create team drops the shared secret on the floor and so it needs to be recalculated from Ctext from team/get with the user's perUserEncryptionKey, which requires full puk chain, which only comes from the network.  And changing that seems beyond the scope of this PR.

File: go/teams/box.go (hunk 6 above)
Comment (maxtaco) [inline go/teams/box.go]: OK, I think it's worth it to be explicit. Can you put this struct into `team.avdl`? Also maybe the same for other json templates.

File: go/libkb/constants.go [hunk 7]
```diff
@@ -569,8 +570,13 @@ type PvlUnparsed struct {
...
 const SharedTeamKeyBoxVersion1 = 1
 
-const TeamDHDerivationString = "Keybase-Derived-Team-NaCl-DH-1"
-const TeamEdDSADerivationString = "Keybase-Derived-Team-NaCl-EdDSA-1"
+const (
+	TeamDHDerivationString       = "Keybase-Derived-Team-NaCl-DH-1"
+	TeamEdDSADerivationString    = "Keybase-Derived-Team-NaCl-EdDSA-1"
+	TeamKBFSDerivationString     = "Keybase-Derived-Team-NaCl-KBFS-1"
+	TeamChatDerivationString     = "Keybase-Derived-Team-NaCl-Chat-1"
+	TeamSaltpackDerivationString = "Keybase-Derived-Team-Nacl-Saltpack-1"
```
Comment (patrickxb) [inline go/libkb/constants.go]: fixed

File: go/teams/box.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	maskBytes, err := mask.MaskBytes()
+	if err != nil {
+		return libkb.NaclDHKeyPair{}, err
+	}
+	var secBytes []byte
+	n := libkb.XORBytes(secBytes, derivedSecret(secret, derivationString), maskBytes)
+	if n != 32 {
+		return libkb.NaclDHKeyPair{}, errors.New("invalid derived secret size")
+	}
+	return libkb.MakeNaclDHKeyPairFromSecretBytes(secBytes)
```
Comment (patrickxb) [inline go/teams/box.go]: Fixed.

File: go/teams/get.go (hunk 2 above)
Comment (patrickxb) [inline go/teams/get.go]: rawTeam has `Box TeamBox`, so the unmarshal fills in TeamBox directly, but I guess `reader_key_masks` isn't in `box` as I thought.

File: go/teams/team.go [hunk 9]
```diff
@@ -0,0 +1,298 @@
+package teams
```
Comment (patrickxb) [inline go/teams/team.go]: The type in it is singular (`Team`).

File: go/teams/box.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+		Nonce:          nonce,
+		Receiver:       encKey.GetKID().ToBytes(),
+		Sender:         t.SenderKID.ToBytes(),
+	}
+
+	plaintext, _, err := encKey.Decrypt(nei)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (patrickxb) [inline go/teams/box.go]: done in team.go

File: go/teams/team.go (hunk 9 above)
Comment (maxtaco) [inline go/teams/team.go]: I cringe at `teams/team.go`

File: go/teams/box.go (hunk 6 above)
Comment (patrickxb) [inline go/teams/box.go]: done.

File: go/teams/box.go (hunk 1 above)
Comment (patrickxb) [inline go/teams/box.go]: done.

File: go/teams/member_set.go (hunk 4 above)
Comment (patrickxb) [inline go/teams/member_set.go]: done

File: go/teams/member_set.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+		if err != nil {
+			return nil, err
+		}
+	}
+	return members, nil
+}
+
+func (m *memberSet) loadMember(ctx context.Context, g *libkb.GlobalContext, username string) (member, error) {
+	user, err := libkb.LoadUser(libkb.NewLoadUserByNameArg(g, username))
```
Comment (patrickxb) [inline go/teams/member_set.go]: done

File: go/teams/create.go [hunk 12]
```diff
@@ -204,7 +209,7 @@ type PerTeamSharedSecretBoxes struct {
 	Generation    PerUserSecretGeneration `json:"generation"`
 	EncryptingKid keybase1.KID            `json:"encrypting_kid"`
 	Nonce         string                  `json:"nonce"`
-	Prev          *string                 `json:"prev"`
+	PrevKey       *string                 `json:"prev"`
```
Comment (patrickxb) [inline go/teams/create.go]: I added MaskB64 to teams.avdl, but this code here is for team create, so not going to touch it as part of this PR.

**Comment by patrickxb** [comment]
Ok, PTAL.  Thanks!

File: go/teams/team.go [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+		if err != nil {
+			return nil, err
+		}
+
+		teamKey, err := t.Chain.GetPerTeamKeyAtGeneration(t.Box.Generation)
+		if err != nil {
+			return nil, err
+		}
+
+		if !teamKey.SigKID.Equal(signingKey.GetKID()) {
```
Comment (maxtaco) [inline go/teams/team.go]: This needs to be a constant time comparison (see crypto/subtle)

File: go/teams/teams.go [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+		if !teamKey.SigKID.Equal(signingKey.GetKID()) {
+			return nil, errors.New("derived signing key did not match key in team chain")
+		}
+
+		if !teamKey.EncKID.Equal(encryptionKey.GetKID()) {
+			return nil, errors.New("derived encryption key did not match key in team chain")
+		}
+
+		// TODO: check that t.Box.SenderKID is a known device DH key for the
+		// user that signed the link.
```
Comment (maxtaco) [inline go/teams/teams.go]: can you point at a JIRA issue for this?

**Pull request merged**